membershipextras.php

```diff
@@ -386,3 +409,419 @@ function membershipextras_civicrm_alterContent(&$content, $context, $tplName, &$
 function membershipextras_civicrm_entityTypes(&$entityTypes) {
   return _membershipextras_civix_civicrm_entityTypes($entityTypes);
 }
+
+/**
+ * Implements hook_civicrm_permission().
+ *
+ * @link http://wiki.civicrm.org/confluence/display/CRMDOC/hook_civicrm_permission
+ */
+function membershipextras_civicrm_permission(&$permissions) {
+  $prefix = ts('Members-Only Event') . ': ';
+  $permissions['members only event registration'] = $prefix . ts('Can register for members-only events irrespective of membership status');
+}
+
+/**
+ * Implements hook_civicrm_tabset().
+ *
+ * @link https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_tabset/
+ */
+function membershipextras_civicrm_tabset($tabsetName, &$tabs, $context) {
+  // check if the tabset is 'Manage Event' page
+  if ($tabsetName == 'civicrm/event/manage') {
+    if (empty($context['event_id'])) {
+      return;
+    }
+
+    $eventID = $context['event_id'];
+    $url = CRM_Utils_System::url(
+      'civicrm/event/manage/membersonlyevent',
+      'reset=1&id=' . $eventID . '&action=update&component=event');
+
+    $tab['membersonlyevent'] = array(
+      'title' => ts('Members only event settings'),
+      'link' => $url,
+      'valid' => _membershipextras_membersonlyevent_is_tab_valid($eventID),
+      'active' => TRUE,
+      'current' => FALSE,
+      'class' => 'ajaxForm',
+    );
+
+    //Insert this tab into position 4 (after `Online Registration` tab)
+    $tabs = array_merge(
+      array_slice($tabs, 0, 4),
+      $tab,
+      array_slice($tabs, 4)
+    );
+  }
+}
+
+/* Helper functions for membersonlyevent */
```

Comment:
Yeah, like these kind of "helper" functions we should try to move into appropriate classes.